### PR TITLE
new device API

### PR DESCRIPTION
**Pull request by piotrbartman**
after QubesOS/qubes-core-admin/pull/579

Proposed device events:
sometimes event can be fired twice, we can ignore it
e.g. if device is plugged out and in in short time we can have doubled
detaching and attaching...

- device-list-changed: device-added -> device-added{devclass}

- device-list-changed: device-remove -> device-removed{devclass}

- device-property-changed: property_name: TODO?

- device-assignment-changed: created -> device-assign:{devclass}

- device-assignment-changed: removed -> device-unassign:{devclass}

- device-assignment-changed: attached -> device-attach:{devclass}

- device-assignment-changed: detached -> device-detach:{devclass}

- device-assignment-changed: property-set -> device-assignment-changed:{devclass}

https://github.com/QubesOS/qubes-issues/issues/4626

### CODE REVIEW

File: qubesadmin/devices.py [hunk 1]
```diff
@@ -24,102 +25,885 @@
...
+# e.g. if device is plugged out and in in short time we can have doubled
+# detaching and attaching...
+# - device-list-changed: device-added -> device-added{devclass}
+# - device-list-changed: device-remove -> device-removed{devclass}
+# - device-property-changed: property_name: TODO?
+# - device-assignment-changed: created -> device-assign:{devclass}
+# - device-assignment-changed: removed -> device-unassign:{devclass}
+# - device-assignment-changed: attached -> device-attach:{devclass}
+# - device-assignment-changed: detached -> device-detach:{devclass}
+# - device-assignment-changed: property-set -> device-assignment-changed:{devclass}
```
Comment (marmarek) [inline qubesadmin/devices.py]: Is this comment relevant to this repo? Events are generated in core-admin one. But also, I think most if not all are already done?

File: qubesadmin/tools/qvm_device.py [hunk 2]
```diff
@@ -129,28 +134,97 @@ def attach_device(args):
...
+    options = dict(opt.split('=', 1) for opt in args.option or [])
+    if args.ro:
+        options['read-only'] = 'yes'
+    options['identity'] = device.self_identity
+    if args.port:
+        options['identity'] = 'any'
+    assignment.options = options
+    vm.devices[args.devclass].assign(assignment)
+    if vm.is_running() and not assignment.attached:
+        print("Assigned. Now you can manually attach device or restart domain.")
```
Comment (marmarek) [inline qubesadmin/tools/qvm_device.py]: This (and other below) should respect `-q` (same below).

File: qubesadmin/tools/qvm_device.py [hunk 3]
```diff
@@ -129,28 +134,97 @@ def attach_device(args):
...
+        vm.devices[args.devclass].unassign(assignment)
+        if assignment.attached:
+            print("Unassigned. Now you can manually detach device "
+                  "or restart domain.")
+    else:
+        for device_assignment in (
+                vm.devices[args.devclass].get_assigned_devices()):
+            vm.devices[args.devclass].unassign(device_assignment)
+            if device_assignment.attached:
+                print("Unassigned. Now you can manually detach device "
```
Comment (marmarek) [inline qubesadmin/tools/qvm_device.py]: This (besides `-q`) should also print which device.

File: qubesadmin/tools/qvm_device.py [hunk 4]
```diff
@@ -303,10 +424,4 @@ def main(args=None, app=None):
 
 
 if __name__ == '__main__':
-    # Special treatment for '--list-device-classes' (alias --list-classes)
-    curr_action = sys.argv[1:]
-    if set(curr_action).intersection(
-            {'--list-device-classes', '--list-classes'}):
-        sys.exit(main(args=['', '--list-device-classes']))
-
```
Comment (marmarek) [inline qubesadmin/tools/qvm_device.py:311]: What happened to this one (besides it being already slightly broken due to requiring a device class anyway)?

**Review by marmarek — CHANGES_REQUESTED** [review]

File: qubesadmin/devices.py (hunk 1 above)
Comment (piotrbartman) [reply qubesadmin/devices.py]: youre right

File: qubesadmin/tools/qvm_device.py (hunk 2 above)
Comment (piotrbartman) [reply qubesadmin/tools/qvm_device.py]: done

File: qubesadmin/tools/qvm_device.py (hunk 3 above)
Comment (piotrbartman) [reply qubesadmin/tools/qvm_device.py]: done

File: qubesadmin/tools/qvm_device.py (hunk 4 above)
Comment (piotrbartman) [reply qubesadmin/tools/qvm_device.py:311]: I slightly fix it and now device class is not required :D  (see lines 408-411).

**Comment by marmarek** [comment]
Starting a qube with CDROM seems to be broken: https://openqa.qubes-os.org/tests/96799#step/TC_00_Basic/4

**Comment by marmarek** [comment]
And also, unit tests need an update (and also pylint is unhappy)

**Comment by marmarek** [comment]
I keep device_protocol.py review in https://github.com/QubesOS/qubes-core-admin/pull/579, other changes here looks okay (but I haven't tested the most recent version here yet).

**Comment by marmarek** [comment]
`qvm-start --cdrom` still is not happy:
```
Traceback (most recent call last):
  File "/usr/lib64/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/qubes/tests/integ/basic.py", line 105, in test_120_start_standalone_with_cdrom_dom0
    self.assertEqual(p.returncode, 0, stdout)
AssertionError: 1 != 0 : b'Can not remove an required assignment from a non halted qube.\n'
```

See also updated comment from the today's test run, most failures looks to be related to the devices API rework (but some are probably "just" tests to be updated).

**Review by marmarek — APPROVED** [review]
